src/core/raster/qgsrastercontourrenderer.cpp

```diff
@@ -0,0 +1,226 @@
+  double downscale = elem.attribute( QStringLiteral( "downscale" ), QStringLiteral( "4" ) ).toDouble();
+
+  r->setInputBand( inputBand );
+  r->setContourInterval( contourInterval );
+  r->setContourIndexInterval( contourIndexInterval );
+  r->setDownscale( downscale );
+
+  QDomElement symbolsElem = elem.firstChildElement( QStringLiteral( "symbols" ) );
+  if ( !symbolsElem.isNull() )
+  {
+    QgsSymbolMap symbolMap = QgsSymbolLayerUtils::loadSymbols( symbolsElem, QgsReadWriteContext() );
+    if ( symbolMap.contains( QStringLiteral( "contour" ) ) )
+    {
+      QgsSymbol *symbol = symbolMap.take( QStringLiteral( "contour" ) );
+      if ( symbol->type() == QgsSymbol::Line )
+        r->setContourSymbol( static_cast<QgsLineSymbol *>( symbol ) );
+    }
+    if ( symbolMap.contains( QStringLiteral( "index-contour" ) ) )
+    {
+      QgsSymbol *symbol = symbolMap.take( QStringLiteral( "index-contour" ) );
+      if ( symbol->type() == QgsSymbol::Line )
+        r->setContourIndexSymbol( static_cast<QgsLineSymbol *>( symbol ) );
+    }
+  }
+  return r;
+}
+
+void QgsRasterContourRenderer::writeXml( QDomDocument &doc, QDomElement &parentElem ) const
+{
+  if ( parentElem.isNull() )
+  {
+    return;
+  }
+
+  QDomElement rasterRendererElem = doc.createElement( QStringLiteral( "rasterrenderer" ) );
+  _writeXml( doc, rasterRendererElem );
+
+  rasterRendererElem.setAttribute( QStringLiteral( "band" ), mInputBand );
+  rasterRendererElem.setAttribute( QStringLiteral( "contour-interval" ), mContourInterval );
+  rasterRendererElem.setAttribute( QStringLiteral( "contour-index-interval" ), mContourIndexInterval );
+  rasterRendererElem.setAttribute( QStringLiteral( "downscale" ), mDownscale );
+
+  QgsSymbolMap symbols;
+  symbols[QStringLiteral( "contour" )] = mContourSymbol.get();
+  if ( mContourIndexSymbol )
+    symbols[QStringLiteral( "index-contour" )] = mContourIndexSymbol.get();
+  QDomElement symbolsElem = QgsSymbolLayerUtils::saveSymbols( symbols, QStringLiteral( "symbols" ), doc, QgsReadWriteContext() );
+  rasterRendererElem.appendChild( symbolsElem );
+
+  parentElem.appendChild( rasterRendererElem );
+}
+
+struct ContourWriterData
+{
+  QPainter *painter;
+  double scaleX, scaleY;
+  QgsLineSymbol *symbol;
+  QgsLineSymbol *indexSymbol;
+  double indexInterval;
+  QgsRenderContext *context;
+};
+
+CPLErr _rasterContourWriter( double dfLevel, int nPoints, double *padfX, double *padfY, void *ptr )
+{
+  Q_UNUSED( dfLevel )
+  ContourWriterData *crData = static_cast<ContourWriterData *>( ptr );
+  QPolygonF polygon( nPoints );
+  QPointF *d = polygon.data();
+  for ( int i = 0; i < nPoints; ++i )
+  {
+    d[i] = QPointF( padfX[i] * crData->scaleX, padfY[i] * crData->scaleY );
+  }
+
+  if ( crData->indexSymbol && !qgsDoubleNear( crData->indexInterval, 0 ) && qgsDoubleNear( fmod( dfLevel, crData->indexInterval ), 0 ) )
```

Comment:
@wonder-sk , any chance you could add an expression context variable that'd contain the height the line being drawn represents? That'd open the door to a lot of extra nice possibilities via data-defined properties.

Reply:
@nirvn it won't be much use unless we have access to the wider QgsRenderContext -- that's where the expression context is taken from

Reply:
Adding height as a variable would be possible... As Nyall says you still wouldn't have access to the main render context, only a local one, but maybe that wouldn't stop you from drawing some fancy looking lines :) In any case, I would prefer to keep the scope of this PR limited, so maybe if you'd like to do a follow-up PR to add that? ;-)

Reply:
Sure, if that's not something you're interested in implementing, someone else can step up. My schedule is hyper busy so likely not me. 

That said, you've walked us that far already, thanks :wink: